Add optimization pass to remove bp stores and restores

lockshaw opened the pull request.

ltfish commented:
Good job!

Linting failed :) (https://travis-ci.org/angr/angr/jobs/498982938) Fix linting and I'll click merge!

ltfish commented:
Is it still WIP? It seems pretty complete to me...

lockshaw commented:
> 
> 
> Is it still WIP? It seems pretty complete to me...

I need to add tests

ltfish commented:
Ping? Any progress?

lockshaw commented:
Just added a basic test case and updated it a bit (improving stack pointer tracking made a previously non-functioning part of the decompiler work, causing the bp save to look different). Should be ready to merge in now.

ltfish commented on tests/test_bp_save_optimization.py:
```diff
@@ -0,0 +1,39 @@
+
+def test_bp_save_fauxware(arch):
+    p = angr.Project(
+        test_location + '/{}/fauxware'.format(arch), auto_load_libs=False)
```
Nitpick: Can you use `os.path.join()` instead?

lockshaw replied:
Yup!

ltfish commented:
LGTM. Thanks. Will merge once tests pass.

ltfish commented:
Thanks!
